### PR TITLE
Depth average heat flux

### PR DESCRIPTION
**Pull request by ian-r-rose**
This adds another statistic to the DepthAverage postprocessor which calculates the vertical heat flux through the model.

There are a truly daunting number of tests which use this postprocessor, so I probably won't attempt to update them until I have the go-ahead.

### CODE REVIEW

File: doc/modules/changes.h [hunk 1]
```diff
@@ -7,6 +7,11 @@
  *
  * <ol>
  *
+ * <li> New: The DepthAverage postprocessor now calculates the laterally
```
Comment (bangerth) [inline doc/modules/changes.h]: "now calculates" -> "can now also calculate"

File: include/aspect/simulator_access.h [hunk 2]
```diff
@@ -399,6 +399,18 @@ namespace aspect
...
+       * convention of positive heat flux when it flows upwards.
+       *
+       * @param values The output vector of depth averaged values. The
+       * function takes the pre-existing size of this vector as the number of
+       * depth slices.
+       */
+      void
+      get_depth_average_heat_flux(std::vector<double> &values) const;
+      /** @} */
       /** @} */
```
Comment (bangerth) [inline include/aspect/simulator_access.h]: No need to duplicate the closing end tag.

File: source/simulator/helper_functions.cc [hunk 3]
```diff
@@ -1207,6 +1207,76 @@ namespace aspect
     compute_depth_average(values, f);
   }
 
+  namespace
+  {
+    template <int dim>
+    class FunctorDepthAverageHeatFlux
```
Comment (bangerth) [inline source/simulator/helper_functions.cc]: Here (and maybe in other places), I would actually prefer a name that emphasizes the fact that you compute the *vertical* heat flux. To me, a flux is a vector quantity, but you only compute a scalar quantity -- but the names don't reflect this.

**Comment by bangerth** [comment]
I think the fundamental problem (that these functions are implemented in the simulator, rather than in a plugin) results from the fact that at least a subset of them are called not only from output plugins, but also from the adiabatic initial condition thingy. At the time, we felt that it would be useful to unify the code in one place rather than duplicating it. In hindsight, I can think of better ways of doing it -- e.g., have a "free" class (or a collection of free functions) somewhere that is responsible for computing depth average by accessing a `SimulatorAccess` object, and let these functions be used by all of the places that current call directly into the solver.

That's a much bigger change, of course. I'm ok with going the route suggested here for now, if you rethink the name.

File: source/simulator/helper_functions.cc (hunk 3 above)
Comment (ian-r-rose) [inline source/simulator/helper_functions.cc]: Agreed, I'll change this.

**Comment by bangerth** [comment]
This looks good in principle but there are a few tests that produce different output (rather than simply adding output to the end of the line). Examples are
- diffusion
- depth-dependent-box-file-simple

Can you verify that these produce the same output for you with and without the patch?

**Comment by ian-r-rose** [comment]
I also saw those and was puzzled.  I was kind of hoping to see what the tester did -- I take it it is down right now?

**Comment by bangerth** [comment]
I don't know. The cdash instance only shows results for the past 3 days, so it may simply not have had anything to eat for the past 3 days. @tjhei is in China this week, so we may have to wait for a couple of days.

**Comment by ian-r-rose** [comment]
Okay, I am going to hold off on trying to figure out what is going on until
then, I think...

On Wed, Aug 12, 2015 at 3:22 PM, Wolfgang Bangerth <notifications@github.com

**Comment by jdannberg** [comment]
I think the tester is not working right now and we have to wait until Timo is back in Clemson...

**Comment by bangerth** [comment]
@ian-r-rose : I think the tester is working again. Can you rebase and 'git push -f' again? This should get the tester to run through this patch again, and hopefully provide meaningful output. (Currently the tester shows a change in indentation, but I think that may be mistaken.)

**Comment by ian-r-rose** [comment]
Okay, I have rebased.  I will take a look at tester output as it comes in.

**Comment by bangerth** [comment]
There are still two tests that fail :-(

**Comment by ian-r-rose** [comment]
Yeah, I am not sure what to do about that.  They are not failing by huge amounts...

**Comment by bangerth** [comment]
OK, that's just the results you have on your machines as opposed to what Timo's machine produces. Can you just update the results in the patch so they match Timo's? You'll just have to apply this patch: http://www.math.clemson.edu/~heister/aspect-logs/d0dcf83041088b428f4146211a69b37c9fd8c636/changes.diff

**Comment by ian-r-rose** [comment]
Patch is applied, tests passing.

File: tests/depth-dependent-box-file-simple/depth_average.gnuplot [hunk 4]
```diff
@@ -4,144 +4,144 @@
 #
 # For a description of the GNUPLOT format see the GNUPLOT manual.
 #
-# <x> <y> <temperature> <adiabatic_temperature> <velocity_magnitude> <sinking_velocity> <Vs> <Vp> <viscosity> 
-0 -1e+07 398 0 7.35903e-32 3.16065e-09 -1 -1 1.6e+21 
-300000 -1e+07 398 0 7.35903e-32 3.16065e-09 -1 -1 1.6e+21 
+# <x> <y> <temperature> <adiabatic_temperature> <velocity_magnitude> <sinking_velocity> <Vs> <Vp> <viscosity> <vertical_heat_flux> 
+0 -1e+07 398 0 7.35903e-32 6.30385e-10 -1 -1 1.6e+21 0.00333333 
+300000 -1e+07 398 0 7.35903e-32 6.30385e-10 -1 -1 1.6e+21 0.00333333
```
Comment (bangerth) [inline tests/depth-dependent-box-file-simple/depth_average.gnuplot]: Something is strange here -- the velocity is around 7e-32 but the sinking velocity is around 6e-9. This makes no real sense.

**Comment by ian-r-rose** [comment]
Yes, I agree that something strange is going on there.  I took a look at this, and I get the same result in the master branch.  So I am not even sure how both this PR and master could be passing the tests for this.

**Comment by bangerth** [comment]
So the velocity depth average actually computes the *square* of the velocity:
```
    template <int dim>
    class FunctorDepthAverageVelocityMagnitude
    {
      public:
        FunctorDepthAverageVelocityMagnitude(const FEValuesExtractors::Vector &field)
          : field_(field) {}

        bool need_material_properties() const
        {
          return false;
        }

        void setup(const unsigned int q_points)
        {
          velocity_values.resize(q_points);
        }

        void operator()(const MaterialModel::MaterialModelInputs<dim> &,
                        const MaterialModel::MaterialModelOutputs<dim> &,
                        FEValues<dim> &fe_values,
                        const LinearAlgebra::BlockVector &solution,
                        std::vector<double> &output)
        {
          fe_values[field_].get_function_values (solution, velocity_values);
          for (unsigned int q=0; q<output.size(); ++q)
            output[q] = velocity_values[q] * velocity_values[q];
        }
```
This is probably a bug in itself -- would anyone agree?

Whereas the sinking velocity computes this:
```
    template <int dim>
    class FunctorDepthAverageSinkingVelocity
    {
      public:
        FunctorDepthAverageSinkingVelocity(const FEValuesExtractors::Vector &field, GravityModel::Interface<dim> *gravity)
          : field_(field), gravity_(gravity) {}

        bool need_material_properties() const
        {
          // this is needed because we want to access in.position in operator()
          return true;
        }

        void setup(const unsigned int q_points)
        {
          velocity_values.resize(q_points);
        }

        void operator()(const MaterialModel::MaterialModelInputs<dim> &in,
                        const MaterialModel::MaterialModelOutputs<dim> &,
                        FEValues<dim> &fe_values,
                        const LinearAlgebra::BlockVector &solution,
                        std::vector<double> &output)
        {
          fe_values[field_].get_function_values (solution, velocity_values);
          for (unsigned int q=0; q<output.size(); ++q)
            {
              const Tensor<1,dim> g = gravity_->gravity_vector(in.position[q]);
              const Tensor<1,dim> vertical = (g.norm() > 0 ? g/g.norm() : Tensor<1,dim>());

              output[q] = std::fabs(std::min(0.0, velocity_values[q] * vertical))*year_in_seconds;
            }
        }
```
So it unconditionally multiplies by `year_in_seconds` and therefore yields a large ratio to the velocity itself. That resolves my uncertainty about what is going on. It's still not clear to me, however, why your patch would actually *change* the numbers you compute...

File: tests/depth-dependent-box-file-simple/depth_average.gnuplot (hunk 4 above)
Comment (bangerth) [inline tests/depth-dependent-box-file-simple/depth_average.gnuplot]: The other part of course is: why is the sinking velocity *different* by about a factor of two, while the velocity itself is the same!?

**Comment by bangerth** [comment]
You mean you get the different sinking velocities also on your master?

**Comment by ian-r-rose** [comment]
I agree that it looks like a bug in both of those depth average functions.

On my machine, there is no change between the master and this branch for those values, so I am not sure what is happening with the tester.

**Comment by ian-r-rose** [comment]
This is very strange.  I can pass those tests on my machine, but manually diffing shows the same differences we see here.  Perhaps there is a bug in the ctest comparisons?

**Comment by bangerth** [comment]
Ah, I see now -- you probably run numdiff (right?) which has an absolute tolerance (see `tests/cmake/diff_test.sh`) and the previous results were already below it. So the results were different for *all of us*, not just you, but nobody ever noticed it.

So if you can confirm that you are using numdiff, then everything should be A OK.

**Comment by ian-r-rose** [comment]
Yes, I am running numdiff.

**Comment by bangerth** [comment]
OK, great -- I think we figured it out! Many thanks for working through all of these contortions!

**Pull request merged**